### PR TITLE
chore(Dependencies): update major ones

### LINKED ISSUE #20: Migrate to React 15.5
React 16 is coming. Until then the facebook team plans to release React version that will help migrating to it. 

We need to migrate to React 15.5 and fix everything that is going to be deprecated.

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

1. react/react-dom
2. prettier
3. jest/enzyme

**Result**

fixes #20 

1. big changes
    - prop-types is a separate lib
        - examples
        - tests
   - tests still give warnings for some reason
      - https://travis-ci.org/algolia/react-instantsearch/builds/222461802#L315 for example...
      - now fixed in 53a939dac0f7821aade01dd324f602c99719519c
2. cool changes to the code
    - reformat everything

### CODE REVIEW

File: docgen/webpack.config.start.babel.js [hunk 1]
```diff
@@ -3,10 +3,7 @@
 import webpack from 'webpack';
 import webpackConfig from './webpack.config.babel.js';
 
-const {
-  HotModuleReplacementPlugin,
-  NoEmitOnErrorsPlugin,
-} = webpack;
+const { HotModuleReplacementPlugin, NoEmitOnErrorsPlugin } = webpack;
```
Comment (Haroenv) [inline docgen/webpack.config.start.babel.js]: destructuring was using same logic as objects before https://github.com/prettier/prettier/pull/981

File: packages/react-instantsearch/examples/autocomplete/src/App-Hits-And-Facets.js [hunk 2]
```diff
@@ -47,20 +48,17 @@ const connectAutoComplete = createConnector({
...
-            .slice(0, 3)
-            .map(v => ({
-              name: v.name,
-            }));
-        }
-        return acc;
-      },
-      {}
-    );
+    const facets = props.attributes.reduce((acc, attributeName) => {
```
Comment (Haroenv) [inline packages/react-instantsearch/examples/autocomplete/src/App-Hits-And-Facets.js]: https://github.com/prettier/prettier/pull/947

File: packages/react-instantsearch/src/components/Highlighter.js [hunk 3]
```diff
@@ -1,8 +1,13 @@
+import PropTypes from 'prop-types';
 import React from 'react';
 
-export default function Highlighter(
-  { hit, attributeName, highlight, highlightProperty, tagName }
-) {
+export default function Highlighter({
```
Comment (Haroenv) [inline packages/react-instantsearch/src/components/Highlighter.js]: https://github.com/prettier/prettier/pull/1022

**Comment by Haroenv** [comment]
I did it with a codemod everywhere in `src`, but by hand in the examples. Where exactly do you want a `createReactClass`, in the tests?

**Comment by mthuret** [comment]
There are some `createReactClass` in `docgen` and `tests`, I think we should migrate those right?

**Comment by Haroenv** [comment]
All those issues should now be fixed.

File: packages/react-instantsearch/examples/geo-search/README.md [hunk 4]
```diff
@@ -3,7 +3,7 @@ This example shows how to perform a geo search using `react-instantsearch`.
 To start the example:
 
 ```sh
-yarn install --no-lockfile
+yarn install
```
Comment (mthuret) [inline packages/react-instantsearch/examples/geo-search/README.md]: why removing the --no-lockfile?

File: packages/react-instantsearch/scripts/build.sh [hunk 5]
```diff
@@ -37,5 +37,5 @@ do
     -o "${dist_file_min}"
 
   gzip_size=$(gzip -9 < "$dist_file_min" | wc -c | pretty-bytes)
-  echo "=> $dist_file_min gzipped will weight $gzip_size"
+  echo "=> $dist_file_min gzipped will weigh $gzip_size"
```
Comment (mthuret) [inline packages/react-instantsearch/scripts/build.sh]: Unfortunate typo?

Comment (Haroenv) [inline packages/react-instantsearch/scripts/build.sh]: It's one that's on lots of places, so I change them one by one 😄

File: packages/react-instantsearch/examples/geo-search/README.md (hunk 4 above)
Comment (Haroenv) [inline packages/react-instantsearch/examples/geo-search/README.md]: I don't see the value of writing `--no-lockfile` here.

Comment (mthuret) [inline packages/react-instantsearch/examples/geo-search/README.md]: It's there to not create a lock file when checkouting and running the examples. (and its present in every examples).

No matter what, this is not related to the dependencies upgrade.
